rfd/0180-sso-as-mfa-method.md

```diff
@@ -255,197 +403,322 @@ invariant, device trust must be enforced within the SSO MFA check.
-+  ALLOW_AS_MFA_METHOD_ONLY = 3;
-+}
+**SSOResponse**
 
-message OIDCConnectorSpecV3 {
-  ...
-+  // AllowAsMFAMethod represents whether this auth connector can be used as an MFA
-+  // method or not.
-+  AllowAsMFAMethod AllowAsMFAMethod = 19 [(gogoproto.jsontag) = "allow_as_mfa_method,omitempty"];
+```proto
+// SSOResponse is a response to SSOChallenge.
+message SSOResponse {
+  // RequestId is the ID of an SSO auth request.
+  string request_id = 1;
+  // Token is a secret token used to verify the user's SSO MFA session.
+  string token = 2;
 }
+```
 
-message SAMLConnectorSpecV2 {
+```diff
+message MFAAuthenticateChallenge {
   ...
-+  // AllowAsMFAMethod represents whether this auth connector can be used as an MFA
-+  // method or not.
-+  AllowAsMFAMethod AllowAsMFAMethod = 17 [(gogoproto.jsontag) = "allow_as_mfa_method,omitempty"];
++ // SSO Challenge is an sso MFA challenge. If set, the client can go to the
++ // IdP redirect URL to perform an MFA check in the IdP and obtain an MFA
++ // token. This token and sso request pair can then be used as MFA verification.
++ SSOChallenge SSOChallenge = 5;
 }
- 
-message GithubConnectorSpecV3 {
-  ...
-+  // AllowAsMFAMethod represents whether this auth connector can be used as an MFA
-+  // method or not.
-+  AllowAsMFAMethod AllowAsMFAMethod = 10 [(gogoproto.jsontag) = "allow_as_mfa_method,omitempty"];
+```
+
+**SSOMFADevice**
+
+```proto
+// SSOMFADevice contains details of an SSO MFA method.
+message SSOMFADevice {
+  // connector_id is the ID of the SSO connector.
+  string connector_id = 1;
+  // connector_type is the type of the SSO connector.
+  string connector_type = 2;
 }
 ```
 
-**CreatePrivilegedToken**
 ```diff
-message OIDCAuthRequest {
+message MFADevice {
   ...
-+  // CreatePrivilegedToken is an option to create a privileged token instead of creating 
-+  // a user session. Privileged tokens can be used in place of standard MFA verification for
-+  // privileged actions. This action is only allowed if the auth connector is allowed
-+  // to be used as an MFA method and if the user is pre-authenticated (not first time login).
-+  bool CreatePrivilegedToken = 19 [(gogoproto.jsontag) = "create_privileged_token,omitempty"];
+  oneof device {
+    TOTPDevice totp = 8;
+    U2FDevice u2f = 9;
+    WebauthnDevice webauthn = 10;
++   SSOMFADevice sso = 11;
+  }
 }
+```
 
-message SAMLAuthRequest {
-  ...
-+  // CreatePrivilegedToken is an option to create a privileged token instead of creating 
-+  // a user session. Privileged tokens can be used in place of standard MFA verification for
-+  // privileged actions. This action is only allowed if the auth connector is allowed
-+  // to be used as an MFA method and if the user is pre-authenticated (not first time login).
-+  bool CreatePrivilegedToken = 18 [(gogoproto.jsontag) = "create_token,omitempty"];
+**AuthConnectorMFAMode**
+
+```proto
+// AuthConnectorMFAMode specified the MFA mode of an Auth Connector.
+enum AuthConnectorMFAMode {
```

Comment:
Suggestion: Add an "off" mode here, so have a corresponding mode for mfa.enabled=false